### ISSUE TITLE
UI event state machine feedback

### ISSUE THREAD

**Issue by vietnugent** [mobxjs/mobx#977]
I have a
 1. [x] Question:
TLDR; I'm looking examples of conditional rendering based on computed values.

In my map editing app the UI conceptually resembles an online shopping site - as you draw/add items on the map, the app keeps track of your edits in a work-in-progress holding queue ('shopping cart') and then at some point you can submit your edits to the backend ('checkout') .

![mobx-ui](https://cloud.githubusercontent.com/assets/19377299/25567681/2784d3c6-2da8-11e7-8f16-40971dae12c0.png)

I have 2 global observable data stores, an observable.map for domain data (GIS objects) and an observable(UI state object).  The UI state object acts like a simple state machine for displaying/hiding various UI components based on user actions.  For example, to enter drawing/edit mode the code would call store.openEditForm() which is a mobx action to set the UI state to `Open-Edit-Mode`.  My  EditForm component (an observer) would then check for the `computed` value of current ui state variable to see if it needs to show itself:
```
const EditForm = observable(
  class EditForm extends Component {

    render() {
        if (store.uiState.shouldEditFormOpen) {
          return null;
       }
       return (<Form ....>);
    }
  }
);
```

Initially I only had 1 UI component and everything was working fine until this weekend when I introduced an additional modal to show the WIP queue ('the shopping cart') and started seeing this error:
```
Warning: forceUpdate(...): Cannot update during an existing state transition (such as within `render` or another component's constructor). Render methods should be a pure function of props and state; constructor side-effects are an anti-pattern, but can be moved to `componentWillMount`.
```
The global ui state is set correctly but somehow conditional rendering is no longer working.  If I click on the WIP button the EntryForm shows up instead - For some reason EntryForm.render() still responds to the old state.

I choose this event dispatch approach because 1) it's simple to trigger the EditForm from multiple places, from the NavBar or the welcome screen. 2) Coming from Java background the observable patterns feel familiar.  But as I add more code it feels like I'm reinventing what navigation and redux are trying to do. Any feedbacks are greatly appreciated.

**Comment by vietnugent** [comment]
An update after additional debugging:  I left out a crucial piece of information in the WIP component:render() that is the cause of incorrect state transition and 2 `forceUpdate()` warning messages.  There are 2 observables in render(), the ui visibility state and the domain model.  For whatever reason render() is being called twice.

```javascript
const WIP = observable(
  class WIP extends Component {

    render() {
        if (store.uiState.shouldWIPOpen) {
          return null;
       }
       const list = store.wip.map( ); // build list - this is the problematic line
       return (
      <ul>
       {list}
      </ul>
      );
    }
  }
);
```

**Comment by suprise** [comment]
Can you show more details in the function`store.wip.map( )`? May be in this method you trigger the store update.

**Comment by mweststrate** [comment]
@vietnugent does this answer (second question) also apply to your case? https://github.com/mobxjs/mobx-react#faq

**Comment by vietnugent** [comment]
@suprise: I was able to track down the problem.   Yup one of the states were being updated in component in `map(...)`
